aea/connections/stub/connection.py

```diff
@@ -145,15 +168,47 @@ def __init__(
 
     def read_envelopes(self) -> None:
         """Receive new envelopes, if any."""
-        line = self.input_file.readline()
-        logger.debug("read line: {!r}".format(line))
-        lines = b""
-        while len(line) > 0:
-            lines += line
-            line = self.input_file.readline()
-        if lines != b"":
-            self._process_line(lines)
+        ok = _lock_file(self.input_file)
+        if not ok:
+            # TOFIX(LR) how to handle locking errors
+            return
+
+        lines = self.input_file.readlines()
+        if len(lines) > 0:
             self.input_file.truncate(0)
+            self.input_file.seek(0)
+
+        ok = _unlock_file(self.input_file)
+        # TOFIX(LR) how to handle locking errors
+
+        #
+        if len(lines) == 0:
+            return
+
+        # split lines based on separator
+        content = b"".join(lines)
+        separator = SEPARATOR
+        splits = content.split(separator)
+
+        message = b""
+        parts = 0
+        for split in splits:
+            parts += 1
+            if parts == 5:
+                logger.debug("processing: {!r}".format(message))
+                self._process_line(message)
+                parts = 1
+                message = b""
+                if split == b"\n" or split == b"":
+                    parts = 0
+                    continue
+            message += split + separator
+            if parts == 1:
+                message = message.decode("utf-8").strip().encode("utf-8")
+
+        if 5 > parts > 0:
+            logger.debug("processing: {!r}".format(message))
+            self._process_line(message)
```

Comment:
the fact we use `\n` as message separator is a bit unfortunate since it can occur within the message fields.

Is there a way to improve the parsing? Maybe instead of `readlines()` we could use `read()` and then split with `separator`. 

Please take the following code as a suggestion or a source of inspiration. No action required, I don't want to add non-necessary work:

```python
# assuming 'content' is the content of the file (notice: it's not readlines()):
#    content = input_file.read()  
# The following is a regex over bytes: ".*,.*,.*,.*,"
regex = re.compile((b".*" + SEPARATOR) * 4, re.DOTALL)
iterator = re.finditer(regex, content)
for match in iterator:
    # remove trailing comma?
    message = match.group(0)[:-1]
    logger.debug("processing: {!r}".format(message))
    self._process_line(message)

```

Reply:
Absolutely we can.
Indeed, if we use `,` as a separator the parsing will be straight forward.
Though, I would rather say that the issue here is that we have to allow adding a trailing `\n` as we want to support people `echo`ing to the input file.
Right, regex is the implementation to go with.